plugins/modules/source_control/gitlab/gitlab_group_members.py

```diff
@@ -0,0 +1,228 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2020, Zainab Alsaffar <Zainab.Alsaffar@mail.rit.edu>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+---
+module: gitlab_group_members
+short_description: Manage group members on GitLab Server
+description:
+    - This module allows to add and remove members to/from a group, or change a member's access level in a group on GitLab.
+version_added: '1.2.0'
+author: Zainab Alsaffar (@zanssa)
+requirements:
+    - python-gitlab python module <= 1.15.0
+    - administrator rights on the GitLab server
+options:
+    server_url:
```

Comment:
Sorry that I only noticed that now, but all other GitLab modules extend the auth_basic docs fragment, and thus are using `api_url` instead of `server_url`. Can you adjust the module to also extend that docs fragment (`extends_documentation_fragment: community.general.auth_basic`), and use `from ansible_collections.community.general.plugins.module_utils.gitlab import gitlabAuthentication` to create the `GitLab` object?

Reply:
hmm! I could change it from server_url to api_url, but I would need more clarification on this.

I have went over gitlab_group, gitlab_runner and gitlab_user modules and I noticed that api_url in the examples section, but I don't see this parameter defined with the other Ansible arguments.

Also, when gitlab_instance is created in all modules, I see that the `module` is passed to `gitlabAuthentication` and the way the arguments are defined in those modules is slightly different from what I did as they used `basic_auth_argument_spec` from `from ansible.module_utils.api import basic_auth_argument_spec`

So, I don't think there would be a need for api_url in the examples sections in those modules as I don't see that it has been defined with the other arguments.

Will also this mean I would need to add that utility and make other changes in arguments? 

Another thing that I noticed is that handling check mode was in the functions/methods in those modules not in the calling code. Has this approach been changed? Is it documented in some article that I could use as a reference?

Reply:
The parameter's documentation is added indirectly in https://github.com/ansible-collections/community.general/blob/main/plugins/modules/source_control/gitlab/gitlab_group.py#L24-L25, the values come from https://github.com/ansible-collections/community.general/blob/main/plugins/doc_fragments/auth_basic.py

In argument_spec, the values are added in https://github.com/ansible-collections/community.general/blob/main/plugins/modules/source_control/gitlab/gitlab_group.py#L248 - the source for this is https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/api.py#L55-L64.

gitlabAuthentication is defined in https://github.com/ansible-collections/community.general/blob/main/plugins/module_utils/gitlab.py#L79-L106 which uses the options defined in the above code to connect to the correct GitLab instance with the correct connection parameters (api_url, api_user/api_password or api_token, and validate_certs).

> Another thing that I noticed is that handling check mode was in the functions/methods in those modules not in the calling code. Has this approach been changed? Is it documented in some article that I could use as a reference?

That's probably more a style issue. IMO either there's complex logic (which at some point needs to know about check mode) in the called function, or otherwise (if it is just a `if check_mode: return` in the beginning) it's better if that is done on the caller's side. But that's something that can be debated.

But then I think it is bad style to exit the module (without failure) in case of check mode inside the called function, but do the regular module exit (without check mode) *outside* the function. That is something I see in existing gitlab_* modules, and that's something I'd really change if I would maintain them, for sake of making it easier to follow the control flow.

Reply:
I see. Thanks for clarifying that.
It should be all set now.

You may have an answer for this:

I have pylint enabled in my visual studio and it showed this message: `Unable to import 'ansible_collections.community.general.plugins.module_utils.gitlab' pylint(import-error)` when I typed the import sentence although I had the collection installed into one of the values for PYTHONPATH. Also, I have tested the module and it is working fine. So, I don't know why I am getting this message. I looked online about that but I haven't found an explanation.